addons/binding/org.openhab.binding.rotelra1x/src/main/java/org/openhab/binding/rotelra1x/handler/RotelRa1xHandler.java

```diff
@@ -0,0 +1,327 @@
+                                // be updated. (making run() and initialize() synchronized doesn't work)
+        } catch (InterruptedException e1) {
+            return;
+        }
+        while (connected && !exit) {
+            try {
+                String command = readCommand();
+                if (command.equals("volume")) {
+                    PercentType vol = readVolume();
+                    updateState(getThing().getChannel("volume").getUID(), vol);
+                } else if (command.equals("mute")) {
+                    String muteState = readUntil('!');
+                    updateState(getThing().getChannel("mute").getUID(),
+                            "on".equals(muteState) ? OnOffType.ON : OnOffType.OFF);
+                } else if (command.equals("power_off")) {
+                    power = false;
+                    updateState(getThing().getChannel("mute").getUID(), OnOffType.OFF);
+                    updateState(getThing().getChannel("power").getUID(), OnOffType.OFF);
+                    updateState(getThing().getChannel("volume").getUID(), UnDefType.NULL);
+                    updateState(getThing().getChannel("source").getUID(), UnDefType.NULL);
+                } else if (command.equals("power_on")) {
+                    power = true;
+                    updateState(getThing().getChannel("power").getUID(), OnOffType.ON);
+                    powerOnRefresh();
+                } else if (command.equals("power")) {
+                    String state = readUntil('!');
+                    if (state.equals("on")) {
+                        power = true;
+                        updateState(getThing().getChannel("power").getUID(), OnOffType.ON);
+                        powerOnRefresh();
+                    } else if (state.equals("standby")) {
+                        updateState(getThing().getChannel("mute").getUID(), OnOffType.OFF);
+                        updateState(getThing().getChannel("power").getUID(), OnOffType.OFF);
+                        power = false;
+                    }
+                } else if (command.equals("dimmer")) {
+                    updateState(getThing().getChannel("dimmer").getUID(), readDimmer());
+                } else if (command.equals("freq")) {
+                    updateState(getThing().getChannel("frequency").getUID(), readFrequency());
+                } else if (command.equals("source")) {
+                    updateState(getThing().getChannel("source").getUID(), new StringType(readUntil('!')));
+                } else if (command.equals("display")) {
+                    String str_length = readUntil(',');
+                    int length = Integer.parseInt(str_length);
+                    byte[] data = new byte[length];
+                    int off = 0;
+                    while (off != length) {
+                        int r = serialPort.getInputStream().read(data, off, length - off);
```

Comment:
What is `r`? Please do not use single letter variable names, except for `i` and `e`. I think you should make an external method to read data with a certain length.